### ISSUE TITLE
Comparison with `Literal` `int`, `Literal` `Enum` and `frozenset` occupancy constrains value differently and sometimes incorrectly

### ISSUE THREAD

**Issue by alwaysmpe** [microsoft/pyright#8641]
**Describe the bug**

I have an integer enum and a set of those enum values. Depending on how a value is compared to the enum and how constructs are typed I get inconsistent behaviour.

*  If I do a direct comparison with the enum value in a branch, pyright is able to correctly deduce that the value in the branch is constrained to the enum, however doesn't then raise an unnecessary comparison error if a second comparison is done with the underlying value.
* If I have a frozenset of values typed as containing the enum type but not literal values, it is constrained to that type and further comparisons to specific values do not raise an error.
* If I have a frozenset typed as containing literal values, further comparisons to one of the constrained literal values raises an unnecessary comparison error.

**Code or Screenshots**

```python
from enum import Enum
from typing import Literal

class IntVal(int, Enum):
    one = 1
    two = 2
    three = 3

type OneTwo = Literal[
    IntVal.one,
    IntVal.two,
]

one_two_set: frozenset[OneTwo] = frozenset({
    IntVal.one,
    IntVal.two,
})

int_val_set: frozenset[IntVal] = frozenset({
    IntVal.one,
    IntVal.two,
})

def foo(t: int):
    if t == IntVal.one:
        # The below does not report an unnecessary comparison but perhaps should
        if t == 1:
            pass
    if t in one_two_set:
        if t == IntVal.one:
            # The below reports an unnecessary comparison, incorrectly claiming it will never be true
            if t == 1:
                pass
    if t in int_val_set:
        # The below does not report an unnecessary comparison, unlike the above branch
        if t == 1:
            pass
```

The message I get:
> Condition will always evaluate to False since the types "Literal[IntVal.one, IntVal.two]" and "Literal[1]" have no overlap  (reportUnnecessaryComparison)

**VS Code extension or command-line**

pylance, pyright cli and playground

**Comment by erictraut** [comment]
There are four separate issues that I think you're reporting here:
1. Pyright is not honoring `Literal[IntVal.one]` as a subtype of `Literal[1]`. This leads to a false positive `reportUnnecessaryComparison` error.
2. Pyright doesn't narrow `t` from `int` to `Literal[IntVal.one]` after a `t == IntVal.one` conditional expression.
3. Pyright does not narrow `t` from `IntVal` to `Literal[IntVal.one]` after a `t == 1` conditional expression.
4. Pyright doesn't reporting a `reportUnnecessaryComparison` error for `if t == 1:` when `t` is know to be of type `IntVal.one`

Of these, I consider the first one to be the most important because the correct behavior is mandated by type theory and the Python typing spec, and pyright is violating this. I consider this a bug, and it will be fixed in the next release.

The other three are not covered by the typing spec. I consider those enhancement requests. I'm not convinced these cases are common enough to merit adding special-case logic in pyright. For that reason, I'm going to reject these enhancement requests at this time. I'm willing to revisit this decision in the future based on feedback from additional pyright users.

For comparison, here's how mypy and pyright (with the above bug fix in place) handle various narrowing cases. Pyright's narrowing is correct in all cases (i.e. it doesn't produce incorrect narrowing results) but is not ideal (i.e. it doesn't always produce the narrowest answer). Mypy's narrowing is both incorrect and incomplete in places.

```python
def foo(t: int) -> None:
    if t == IntVal.one:
        # pyright reveals "int" (correct but not ideal)
        # mypy reveals "Literal[IntVal.one]" (ideal)
        # ideal answer: "Literal[IntVal.one]"
        reveal_type(t)
        if t == 1:
            # pyright reveals "Literal[1]" (correct but not ideal)
            # mypy reveals "Never" (incorrect)
            # ideal answer: "Literal[IntVal.one]"
            reveal_type(t)
            pass
    if t in one_two_set:
        # pyright reveals "Literal[IntVal.one, IntVal.two]" (ideal)
        # mypy reveals "int" (correct but not ideal)
        # ideal answer: "Literal[IntVal.one, IntVal.two]"
        reveal_type(t)
        if t == IntVal.one:
            # pyright reveals "Literal[IntVal.one]" (ideal)
            # mypy reveals "Literal[IntVal.one]" (ideal)
            # ideal answer: "Literal[IntVal.one]"
            reveal_type(t)
            if t == 1:
                # pyright reveals "Literal[IntVal.one]" (ideal)
                # mypy reveals "Never" (incorrect)
                # ideal answer: "Literal[IntVal.one]"
                reveal_type(t)
                pass
    if t in int_val_set:
        # pyright reveals "IntVal" (ideal)
        # mypy reveals "int" (correct but not ideal)
        # ideal answer: "IntVal"
        reveal_type(t)
        if t == 1:
            # pyright reveals "IntVal" (correct but not ideal)
            # mypy reveals "int" (correct but not ideal)
            # ideal answer: "Literal[IntVal.one]"
            reveal_type(t)
            pass

def bar(t: IntVal):
    if t == 1:
        # pyright reveals "IntVal" (correct but not ideal)
        # mypy reveals "Never" (incorrect)
        # ideal answer: "Literal[IntVal.one]"
        reveal_type(t)
    if t == IntVal.one:
        # pyright reveals "Literal[IntVal.one]" (ideal)
        # mypy reveals "Literal[IntVal.one]" (ideal)
        # ideal answer: "Literal[IntVal.one]"
        reveal_type(t)
```

**Comment by alwaysmpe** [comment]
Thanks for responding so quickly, your comments are always insightful.

I agree, for me the false positive is a bigger issue than any of the false negatives, I have no issue with them being left as is. I use Pyright over Mypy specifically due to fewer false positives.

I'd be happier working in C++ if only I didn't have clients and customers.

**Comment by erictraut** [comment]
This is included in pyright 1.1.375.
